wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -310,13 +314,67 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update) throws Conflic
+     * @param update
+     *         new workspace configuration
+     */
+    private void updateWorkspaceSnapshots(String workspaceId, Workspace current, Workspace update) {
+        // this is needed to convert DTO into EnvironmentImpl object, otherwise equals will fail
+        final Map<String, Environment> updateEnvironments = update.getConfig()
+                                                                  .getEnvironments()
+                                                                  .entrySet()
+                                                                  .stream()
+                                                                  .collect(Collectors.toMap(Map.Entry::getKey,
+                                                                                            env -> new EnvironmentImpl(env.getValue())));
+
+        final Map <String, Environment> currentDiff = Sets.difference(current.getConfig().getEnvironments().entrySet(),
+                                                                      updateEnvironments.entrySet())
+                                                          .stream()
+                                                          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        final Map <String, Environment> updateDiff = Sets.difference(updateEnvironments.entrySet(),
+                                                                     current.getConfig().getEnvironments().entrySet())
+                                                         .stream()
+                                                         .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        for (Map.Entry<String, Environment> env : currentDiff.entrySet()) {
+            Environment envValue = env.getValue();
+            if (Collections.frequency(currentDiff.values(), envValue) == 1 &&
+                Collections.frequency(updateDiff.values(), envValue) == 1 &&
+                !updateDiff.containsKey(env.getKey())) {
+                try {
+                    updateEnvironmentSnapshots(workspaceId, env.getKey(), updateDiff.entrySet()
+                                                                                    .stream()
+                                                                                    .filter(e -> e.getValue().equals(envValue))
+                                                                                    .findAny()
+                                                                                    .orElse(null) // should never happens
+                                                                                    .getKey());
+                } catch (ConflictException | SnapshotException | NotFoundException err) {
+                    LOG.error("Failed to update snapshots of '%s' environment in '%s' workspace.", env.getKey(), workspaceId, err);
+                    safeRemoveEnvironmentSnapshots(workspaceId, env.getKey());
```

Comment:
i wouldn't use _safe_ for such kind of method it's usually called called _actionQuietly_ or just _action_.